### DISCUSSION TITLE
PhotoService의 책임은 어디까지 가져갈 수 있을까?

### DISCUSSION THREAD

**Discussion by Kim0914** [Busan-Dinosaur/Foodbowl-Backend#114] [의논해볼까요?]
개발을 하면서 궁금한 점이 생겼는데요 ㅎㅎ

PhotoService의 책임은 어디까지 가져갈 수 있을지 궁금했습니다 !

사진 객체를 생성하는 것과 이미지를 업로드하는 기능은 성격이 다르다고 생각해서 현재는 분리해서 구현했습니다.
`PhotoService`는 Photo 객체의 생명주기를 관리하는 책임을 부여했고, `PhotoUtils`은 파일(이미지)를 업로드 하는 책임을 부여했습니다.

다만, 이런 구조로 구현하다보니 `PhotoService`가 너무 하는 역할이 없지 않나? 라는 생각도 들더라구요 !

관련해서 의견 자유롭게 말씀해주시면 좋을 것 같아요 :)

**Comment by woo-chang** [comment]
사진을 업로드하는 행위는 비즈니스 로직에서 **외부의 서비스를 사용**하는 것이라 생각해요. 그렇기에 파일 업로드 책임은 `PhotoService`가 아닌 다른 객체에 책임을 부여하는 것에 대해서는 적극적으로 동의합니다 👍

사진은 도메인이기보다는 하나의 데이터이기 때문에 도메인 로직이 없어 서비스 계층의 책임이 적어보인다고 생각할 수도 있을 것 같아요. 하지만 사진을 업로드하는 외부 서비스 구현체를 주입받고, 이를 호출하여 경로를 얻은 뒤 DB에 영속화하는 것만으로도 `PhotoService`는 책임을 다하고 있는 것이 아닐까라는 생각이 드네요~

추가적으로 사진 업로드 방식이 바뀐다고 한다면 `PhotoService`에서 참조하고 있는 `PhotoUtils`의 구현체를 수정하는 식으로 변경할 수 있을 것 같아요. 그렇기에 `Utils`라는 네이밍보다는 `PhotoUploader`과 같은 인터페이스를 두고 `PhotoS3Uploader`, `PhotoLocalUploader`과 같이 구현체를 작성할 수 있을 것 같은데 이 부분에 대해서는 어떻게 생각하시나요?! 😙

**Reply by Kim0914** [reply]
오 네이밍 너무 좋은데요? 의견 너무 감사합니다 🙇🏻

`Utils`가 info나 value 같은 안티 느낌이라서 고민이 있었는데 Uploader 좋습니다 ㅎㅎ
